--- 7d04ff7a9efdebb7.copy.metal	2026-10-07 09:14:55
+++ 7d04ff7a9efdebb7.dedupe.metal	2026-10-07 09:38:53
@@ -965,19 +965,19 @@
             long epi_off4 = u.epilogue_operand_base[4];
             epi_off4 += coord_q_cache[q][0] * u.epilogue_operand_strides[4][0];
             epi_off4 += coord_q_cache[q][2] * u.epilogue_operand_strides[4][1];
-            epi_scratch[4] = epi4[epi_off4];
+            epi_scratch[4] = epi_scratch[2];
             long epi_off5 = u.epilogue_operand_base[5];
             epi_off5 += coord_q_cache[q][0] * u.epilogue_operand_strides[5][0];
             epi_off5 += coord_q_cache[q][2] * u.epilogue_operand_strides[5][1];
-            epi_scratch[5] = epi5[epi_off5];
+            epi_scratch[5] = epi_scratch[2];
             long epi_off6 = u.epilogue_operand_base[6];
             epi_off6 += coord_q_cache[q][0] * u.epilogue_operand_strides[6][0];
             epi_off6 += coord_q_cache[q][2] * u.epilogue_operand_strides[6][1];
-            epi_scratch[6] = epi6[epi_off6];
+            epi_scratch[6] = epi_scratch[2];
             long epi_off7 = u.epilogue_operand_base[7];
             epi_off7 += coord_q_cache[q][0] * u.epilogue_operand_strides[7][0];
             epi_off7 += coord_q_cache[q][2] * u.epilogue_operand_strides[7][1];
-            epi_scratch[7] = epi7[epi_off7];
+            epi_scratch[7] = epi_scratch[2];
             long epi_off8 = u.epilogue_operand_base[8];
             epi_off8 += coord_q_cache[q][0] * u.epilogue_operand_strides[8][0];
             epi_off8 += coord_q_cache[q][2] * u.epilogue_operand_strides[8][1];
